avatica/src/main/java/org/apache/calcite/avatica/QueryState.java

```diff
@@ -0,0 +1,474 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.avatica;
+
+import org.apache.calcite.avatica.proto.Common;
+import org.apache.calcite.avatica.proto.Common.MetaDataOperationArgument;
+import org.apache.calcite.avatica.proto.Common.MetaDataOperationArgument.ArgumentType;
+import org.apache.calcite.avatica.remote.MetaDataOperation;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * A struct used to encapsulate the necessary information to reconstitute a ResultSet in the
+ * Avatica server.
+ */
+public class QueryState {
+
+  /**
+   * An enumeration that represents how a ResultSet was created.
+   */
+  public static enum StateType {
+    SQL,
+    METADATA;
+
+    public Common.StateType toProto() {
+      switch (this) {
+      case SQL:
+        return Common.StateType.SQL;
+      case METADATA:
+        return Common.StateType.METADATA;
+      default:
+        return Common.StateType.UNRECOGNIZED;
+      }
+    }
+
+    public static StateType fromProto(Common.StateType protoType) {
+      switch (protoType) {
+      case SQL:
+        return StateType.SQL;
+      case METADATA:
+        return StateType.METADATA;
+      default:
+        throw new IllegalArgumentException("Unhandled StateType " + protoType);
+      }
+    }
+  }
+
+  @JsonProperty("type")
+  public final StateType type;
+
+  @JsonProperty("sql")
+  public final String sql;
+
+  @JsonProperty("metaDataOperation")
+  public final MetaDataOperation metaDataOperation;
+  @JsonProperty("operationArgs")
+  public final Object[] operationArgs;
+
+  /**
+   * Constructor encapsulating a SQL query used to create a result set.
+   * @param sql The SQL query
+   */
+  public QueryState(String sql) {
+    this.sql = Objects.requireNonNull(sql);
+    this.type = StateType.SQL;
+
+    // Null out the members we don't use.
+    this.metaDataOperation = null;
+    this.operationArgs = null;
+  }
+
+  /**
+   * Constructor encapsulating a metadata operation's result set
+   * @param op A pointer to the {@link DatabaseMetaData} operation being invoked
+   * @param args The arguments to the method being invoked
+   */
+  public QueryState(MetaDataOperation op, Object[] args) {
```

Comment:
make this varargs, i.e. `Object... args`? Would save a lot of boilerplate in AvaticaDatabaseMetaData